### PR TITLE
mgr/dashboard: migrate e2e tests to async await

### PR DESCRIPTION
**Pull request by p-se**
Migration of E2E tests from Selenium Promise Manager to native JavaScript async/await

This PR disables the promise manager of Selenium and implements JavaScripts native async/await functionality. The issue about deprecating the Selenium promise manager has been closed on Selenium, also see https://github.com/SeleniumHQ/selenium/issues/2969 for details of the issue and background information.

### CODE REVIEW

**Comment by p-se** [comment]
jenkins retest this please

File: src/pybind/mgr/dashboard/HACKING.rst [hunk 1]
```diff
@@ -184,13 +184,29 @@ Note::
...
 
 The ``PageHelper`` class is supposed to be used for general purpose code that
-can be used on various pages or suites. Examples are
-``getTableCellByContent()``, ``getTabsCount()`` or ``checkCheckbox()``. Every
-method that could be useful on several pages belongs there. Also, methods
+can be used on various pages or suites.
+
+Examples are
+
+- ``getTableCellByContent()`` - jeturns a table cell by its content
```
Comment (votdev) [inline src/pybind/mgr/dashboard/HACKING.rst]:
```suggestion
- ``getTableCellByContent()`` - returns a table cell by its content
```

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/feature-toggles.service.ts [hunk 2]
```diff
@@ -15,11 +15,17 @@ export class FeatureTogglesService {
...
 
-  constructor(private http: HttpClient) {
-    this.featureToggleMap$ = timer(0, this.REFRESH_INTERVAL).pipe(
-      flatMap(() => this.http.get<FeatureTogglesMap>(this.API_URL)),
-      shareReplay(1)
-    );
+  constructor(private http: HttpClient, private zone: NgZone) {
+    this.featureToggleMap$ = this.http.get<FeatureTogglesMap>(this.API_URL).pipe(shareReplay(1));
+    this.zone.runOutsideAngular(() => {
+      window.setInterval(() => {
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/services/feature-toggles.service.ts]: Using RxJS ``timer`` does not need the use of ``NgZone``. Maybe you simply need to add ``delay(0)`` to get it working.

File: src/pybind/mgr/dashboard/frontend/e2e/page-helper.po.ts [hunk 3]
```diff
@@ -45,105 +93,142 @@ export abstract class PageHelper {
...
   }
 
-  getFirstTableCellWithText(content) {
+  getFirstTableCellWithText(content): ElementFinder {
     return element.all(by.cssContainingText('.datatable-body-cell-label', content)).first();
   }
 
   /**
-   * Used for instances where a modal container received the click rather than the desired element.
+   * Used for instances where a modal container would receive the click rather
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/e2e/page-helper.po.ts]: Is this line still related to this method?

File: src/pybind/mgr/dashboard/frontend/e2e/page-helper.po.ts [hunk 4]
```diff
@@ -45,105 +93,142 @@ export abstract class PageHelper {
...
-    return browser
-      .wait(() => {
-        title = $('.panel-title');
-        return title.isPresent();
-      })
-      .then(() => title.getText());
-  }
-
-  getTableCell(content) {
+  // getTitleText() {
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/e2e/page-helper.po.ts]: remove this block?

File: src/pybind/mgr/dashboard/frontend/e2e/pools/pools.po.ts [hunk 5]
```diff
@@ -17,86 +17,69 @@ export class PoolPageHelper extends PageHelper {
...
-            .then(() => {
-              this.setApplications(apps);
-              return element(by.css('cd-submit-button')).click();
-            });
-        });
-    });
+    await nameInput.sendKeys(name);
+    await element(by.cssContainingText('select[name=poolType] option', 'replicated')).click();
+
+    expect(await element(by.css('select[name=poolType] option:checked')).getText()).toBe(
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/e2e/pools/pools.po.ts]: why does some expects have `expect(await element(by...` and others just have `expect(element(by...`

**Comment by votdev** [comment]
```
[14:53:35] E/launcher - Error: TSError: ⨯ Unable to compile TypeScript:
e2e/user-mgmt.po.ts:26:10 - error TS2339: Property 'moveClick' does not exist on type 'UserMgmtPageHelper'.

26     this.moveClick(createButton).then(() => {
            ~~~~~~~~~
e2e/user-mgmt.po.ts:49:10 - error TS2339: Property 'moveClick' does not exist on type 'UserMgmtPageHelper'.

49     this.moveClick(editButton).then(() => {
            ~~~~~~~~~
e2e/user-mgmt.po.ts:83:10 - error TS2339: Property 'moveClick' does not exist on type 'UserMgmtPageHelper'.

83     this.moveClick(createButton).then(() => {
            ~~~~~~~~~
e2e/user-mgmt.po.ts:100:10 - error TS2339: Property 'moveClick' does not exist on type 'UserMgmtPageHelper'.

100     this.moveClick(editButton).then(() => {
             ~~~~~~~~~

    at createTSError (/ceph/src/pybind/mgr/dashboard/frontend/node_modules/ts-node/src/index.ts:245:12)
    at reportTSError (/ceph/src/pybind/mgr/dashboard/frontend/node_modules/ts-node/src/index.ts:249:19)
    at getOutput (/ceph/src/pybind/mgr/dashboard/frontend/node_modules/ts-node/src/index.ts:362:34)
    at Object.compile (/ceph/src/pybind/mgr/dashboard/frontend/node_modules/ts-node/src/index.ts:395:32)
    at Module.m._compile (/ceph/src/pybind/mgr/dashboard/frontend/node_modules/ts-node/src/index.ts:473:43)
    at Module._extensions..js (internal/modules/cjs/loader.js:785:10)
    at Object.require.extensions.<computed> [as .ts] (/ceph/src/pybind/mgr/dashboard/frontend/node_modules/ts-node/src/index.ts:476:12)
    at Module.load (internal/modules/cjs/loader.js:641:32)
    at Function.Module._load (internal/modules/cjs/loader.js:556:12)
    at Module.require (internal/modules/cjs/loader.js:681:19)
```

File: src/pybind/mgr/dashboard/frontend/e2e/pools/pools.po.ts (hunk 5 above)
Comment (p-se) [reply src/pybind/mgr/dashboard/frontend/e2e/pools/pools.po.ts]: Only promises can be awaited. `getText()`, `click()` and `sendKeys()`, for instance, return promises. If an expect does not contain an `await`, then it's either wrong or the value to be tested is not a promise.

**Comment by p-se** [comment]
> ```
> [14:53:35] E/launcher - Error: TSError: ⨯ Unable to compile TypeScript:
> e2e/user-mgmt.po.ts:26:10 - error TS2339: Property 'moveClick' does not exist on type 'UserMgmtPageHelper'.
> 
> 26     this.moveClick(createButton).then(() => {
>             ~~~~~~~~~
> e2e/user-mgmt.po.ts:49:10 - error TS2339: Property 'moveClick' does not exist on type 'UserMgmtPageHelper'.
> 
> 49     this.moveClick(editButton).then(() => {
>             ~~~~~~~~~
> e2e/user-mgmt.po.ts:83:10 - error TS2339: Property 'moveClick' does not exist on type 'UserMgmtPageHelper'.
> 
> 83     this.moveClick(createButton).then(() => {
>             ~~~~~~~~~
> e2e/user-mgmt.po.ts:100:10 - error TS2339: Property 'moveClick' does not exist on type 'UserMgmtPageHelper'.
> 
> 100     this.moveClick(editButton).then(() => {
>              ~~~~~~~~~
> 
>     at createTSError (/ceph/src/pybind/mgr/dashboard/frontend/node_modules/ts-node/src/index.ts:245:12)
>     at reportTSError (/ceph/src/pybind/mgr/dashboard/frontend/node_modules/ts-node/src/index.ts:249:19)
>     at getOutput (/ceph/src/pybind/mgr/dashboard/frontend/node_modules/ts-node/src/index.ts:362:34)
>     at Object.compile (/ceph/src/pybind/mgr/dashboard/frontend/node_modules/ts-node/src/index.ts:395:32)
>     at Module.m._compile (/ceph/src/pybind/mgr/dashboard/frontend/node_modules/ts-node/src/index.ts:473:43)
>     at Module._extensions..js (internal/modules/cjs/loader.js:785:10)
>     at Object.require.extensions.<computed> [as .ts] (/ceph/src/pybind/mgr/dashboard/frontend/node_modules/ts-node/src/index.ts:476:12)
>     at Module.load (internal/modules/cjs/loader.js:641:32)
>     at Function.Module._load (internal/modules/cjs/loader.js:556:12)
>     at Module.require (internal/modules/cjs/loader.js:681:19)
> ```

https://github.com/ceph/ceph/pull/29641

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/feature-toggles.service.ts (hunk 2 above)
Comment (p-se) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/feature-toggles.service.ts]: I'm not sure if I understand it. The timer works fine inside the Angular zone, but I've changed it to run outside of the Angular zone to prevent Protractor from waiting endlessly, as a timer is a recurring event that won't be finished at some point.

File: src/pybind/mgr/dashboard/frontend/e2e/page-helper.po.ts (hunk 4 above)
Comment (p-se) [reply src/pybind/mgr/dashboard/frontend/e2e/page-helper.po.ts]: I commented it out when I started and it seems it wasn't used anywhere, so yes, I'll remove it.

File: src/pybind/mgr/dashboard/frontend/e2e/page-helper.po.ts (hunk 3 above)
Comment (p-se) [reply src/pybind/mgr/dashboard/frontend/e2e/page-helper.po.ts]: I don't think so, I'll remove it.

**Comment by tspmelo** [comment]
FYI https://github.com/ceph/ceph/pull/29779 was merged

File: src/pybind/mgr/dashboard/frontend/e2e/pools/pools.po.ts (hunk 5 above)
Comment (tspmelo) [reply src/pybind/mgr/dashboard/frontend/e2e/pools/pools.po.ts]: Since the `expect` without the inside `await` is currently passing on the tests, I don't think it's necessary to use it there.
But it seems Angular team uses the `await` before the `expect` [[example](https://github.com/angular/protractor/blob/5.4.1/spec/ts/basic/element_spec.ts)], so I would suggesting doing the same:

```suggestion
    await expect(element(by.css('select[name=poolType] option:checked')).getText()).toBe(    
```
